Part 3 of Joe TNS post (#3610)
pulumi/pulumi-hugo: themes/default/content/blog/going-beyond-with-advanced-infrastructure-as-code-use-cases/index.md

---
title: "Going Beyond With Advanced Infrastructure as Code Use Cases"
date: 2023-11-21T03:53:33Z
canonical_url: https://thenewstack.io/going-beyond-with-advanced-infrastructure-as-code-use-cases/
draft: false
meta_desc: "Explore advanced Infrastructure as Code (IaC) with Pulumi, covering cloud development, testing, policy enforcement, and the Automation API in a concise guide."
meta_image: meta.png
authors:
    - joe-duffy
tags:
    - infrastructure-as-code
    - fundamentals

---

*This is the third of a three-part series originally published on The New Stack. Read [Part 1](/blog/infrastructure-as-code-in-any-programming-language/) and [Part 2](/blog/walkthrough-of-adopting-infrastructure-as-code/).*

Engineers who modernize their [Infrastructure as Code](/what-is/what-is-infrastructure-as-code/) with Pulumi get two classes of benefits:

1. Infrastructure as Code to develop cloud infrastructure with code.
2. Pulumi Cloud, which tames cloud infrastructure management at [scale](https://roadmap.sh/videos/scaling-the-unscalable).

We’ve covered a fair bit of the first above, but have yet to scratch the surface for the second.

> Many Infrastructure as Code solutions require that you explicitly manage so-called “state.” This state is an artifact produced that keeps track of what you think your infrastructure looks like, as defined by your Infrastructure as Code, so that it can be easily compared to what your actual infrastructure looks like. This is how diffs and updates can be done. Every Infrastructure as Code tool stores this infrastructure state, which is really just metadata about all the cloud resources, properties and dependencies. However, how much the tool exposes you to it varies.

Reviewer: ```suggestion
Many Infrastructure as Code solutions require that you explicitly manage something typically referred to as “state.” This state is an artifact produced that keeps track of what you think your infrastructure looks like, as defined by your Infrastructure as Code, so that it can be easily compared to what your actual infrastructure looks like. This is how diffs and updates can be done. Every Infrastructure as Code tool stores this infrastructure state, which is really just metadata about all the cloud resources, properties and dependencies. However, how much the tool exposes you to it varies.
```

Managed services like AWS CloudFormation and Azure Resource Manager (ARM) abstract this state away behind a service. Other tools like Terraform require you to explicitly manage that state by hand, often storing it in an S3 bucket or equivalent on your chosen cloud. The managed service approach tends to be easier to use and also avoids common pitfalls including security ones ([like accidentally committing secrets to your Terraform state file in plain text](https://sysdig.com/blog/cloud-breach-terraform-data-theft/)) as well as reliability ones ([like a code merge gone wrong that takes out your infrastructure](https://www.youtube.com/watch?v=ix0Tw8uinWs)). Pulumi Cloud offers both options, but the default is to use it like a managed service.

> The Pulumi Cloud is available for unlimited free usage, has a free tier for teams and offers advanced editions for larger teams and enterprises. Pulumi’s SDK is fully open source and can be used without the service (read more about that here), but the cloud service is the easiest, most secure and most reliable way to use Pulumi Cloud.

Reviewer: ```suggestion
Pulumi Cloud is available for unlimited free usage, has a free tier for teams and offers advanced editions for larger teams and enterprises. Pulumi’s SDK is fully open source and can be used without the service (read more about that here), but the cloud service is the easiest, most secure and most reliable way to use Pulumi Cloud.
```

Reviewer: Is there supposed to be a link associated with "read more about that here"?

> Pulumi Cloud is available for unlimited free usage, has a free tier for teams and offers advanced editions for larger teams and enterprises. Pulumi’s SDK is fully open source and can be used without the service (read more about that here), but the cloud service is the easiest, most secure and most reliable way to use Pulumi Cloud.

Reviewer 2: ```suggestion
Pulumi Cloud is available for unlimited free usage, has a free tier for teams and offers advanced editions for larger teams and enterprises. Pulumi’s SDK is fully open source and can be used without the service (read more about that here), but the cloud service is the easiest, most secure and most reliable way to use Pulumi Cloud.
```
```suggestion
Pulumi Cloud is available for unlimited free usage, has a free tier for teams and offers advanced editions for larger teams and enterprises. Pulumi’s SDK is fully open source and can be used without the service, but the cloud service is the easiest, most secure and most reliable way to use Pulumi Cloud.
```

## Learn More about Infrastructure as Code

The Pulumi [Getting Started guide](/docs/get-started/) is the easiest way to get up and running with Infrastructure as Code. In addition to that, plus exploring the Pulumi Templates and Examples, we recommend some additional resources:

The tutorial series, [Infrastructure as Code Recommended Practices](/blog/iac-recommended-practices-code-organization-and-stacks/), is a great overview of many concerns you will encounter along your Infrastructure as Code journey. This ranges from how to organize your project code and stacks, how to do local development and testing of your code, how to evolve these structures as your needs change, automation scenarios and more.

> The [Using Pulumi](/docs/using-pulumi/) page is a collection of tutorials and getting started guides on specific Infrastructure as Code topics, such as deeper dives into fundamental concepts, how to simplify your infrastructure projects, and more.

Reviewer 2: ```suggestion
The [Using Pulumi](/docs/using-pulumi/) page is a collection of tutorials and getting started guides on specific infrastructure as code topics, such as deeper dives into fundamental concepts, how to simplify your infrastructure projects, and more.
```

> The [Pulumi Learn](/learn/) series is a collection of walkthroughs on specific Infrastructure as Code topics, such as deeper dives into fundamental concepts, how to use abstraction and encapsulation to simplify your infrastructure projects, and more.

Reviewer 2: ```suggestion
The [Using Pulumi](/docs/using-pulumi/) page is a collection of tutorials and getting started guides on specific Infrastructure as Code topics, such as deeper dives into fundamental concepts, how to simplify your infrastructure projects, and more.
```
